### ISSUE TITLE
asyncio.ProactorEventLoop mishandles signal wakeup file descriptor

### ISSUE THREAD

**Issue by cd8a9885-2bba-49b3-8952-f6d83fa766a7** [python/cpython#87079]
BPO | [42913](https://bugs.python.org/issue42913)
--- | :---
Nosy | @gvanrossum, @pfmoore, @tjguk, @asvetlov, @zware, @1st1, @zooba, @hidmic

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-01-12.21:43:53.586>
labels = ['type-bug', '3.8', 'OS-windows', 'expert-asyncio']
title = 'asyncio.ProactorEventLoop mishandles signal wakeup file descriptor'
updated_at = <Date 2021-07-29.05:27:36.588>
user = 'https://github.com/hidmic'
```

bugs.python.org fields:
```python
activity = <Date 2021-07-29.05:27:36.588>
actor = 'gvanrossum'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Windows', 'asyncio']
creation = <Date 2021-01-12.21:43:53.586>
creator = 'hidmic'
dependencies = []
files = []
hgrepos = []
issue_num = 42913
keywords = []
message_count = 6.0
messages = ['384979', '384998', '385198', '385211', '398445', '398457']
nosy_count = 8.0
nosy_names = ['gvanrossum', 'paul.moore', 'tim.golden', 'asvetlov', 'zach.ware', 'yselivanov', 'steve.dower', 'hidmic']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue42913'
versions = ['Python 3.8']
```

</p></details>

**Comment by kumaraditya303** [comment]
What is the use case of creating two loops and then deleting them? When `asyncio` is used on windows, signals cannot be used because windows does not supports them. `asyncio` uses `self-pipe` to wake event loop from blocking calls of `select`. The reproducer makes no sense in real code as usually you  don't have instantiate loop object.

**Comment by gvanrossum** [comment]
IIUC the Windows libc emulation does support some limited signal operations (else how would the code work at all?). Presumably the demo program is meant to represent a more realistic scenario where this bug was discovered by the OP.

That said, I don't recall any of the details so maybe someone else can help decide how this could be done better (maybe @zooba?).

**Comment by kumaraditya303** [comment]
> Presumably the demo program is meant to represent a more realistic scenario where this bug was discovered by the OP.

The bug can only happen if you manually install a wakeup fd or you are creating two loops on the same thread both of which are unsupported by `asyncio`.

**Comment by gvanrossum** [comment]
From the wording of the initial comment I think the OP had a manually installed wakeup fd that was getting clobbered by asyncio. I think it's reasonable to see if there's a way asyncio can somehow support this scenario better.

**Comment by kumaraditya303** [comment]
> I think it's reasonable to see if there's a way asyncio can somehow support this scenario better.

The only way I can think of to support this would be to first allow multiple signal wakeup handlers as currently only 1 can be added. I expect the change to be non trivial.

**Comment by gvanrossum** [comment]
Is there a way to ask whether a fd has a signal handler? Is an asyncio loop without a wakeup fd totally crippled, or merely slightly less performant? I am wondering if an async loop could detect whether a signal wakeup fd is already in use, and then refrain from setting up another one. (And I am reluctant to read the corresponding source code. :-) I don't think changing the `signal` module to support multiple handlers is the way to go, so we agree there.

**Comment by kumaraditya303** [comment]
> Is there a way to ask whether a fd has a signal handler?

Not that I am aware of.

>  Is an asyncio loop without a wakeup fd totally crippled, or merely slightly less performant?

It doesn't matter for performance but only for signal handling which on Linux include subprocesses which use `SIGCHILD`, so subprocess would be broken. 

> I am wondering if an async loop could detect whether a signal wakeup fd is already in use, and then refrain from setting up another one.

It would break subprocesses on Linux using child watcher which use `SIGCHILD`, and there is no API to get the wakeup fd, it can only be set or removed with `-1`.

**Comment by gvanrossum** [comment]
So maybe the minimal solution would be to add something to `signal` to ask for the wakeup fd (if the OS allows it?) and then if a wakeup fd is already set, refrain from clobbering it, and instead disable the subprocess functionality of the event loop. Note that some folks are debating to get rid of the current child watcher machinery (https://github.com/python/cpython/issues/82772).

**Comment by kumaraditya303** [comment]
> if a wakeup fd is already set, refrain from clobbering it, and instead disable the subprocess functionality of the event loop

That seems rather complicated for little gain, even if the entire child watcher machinery ought to be removed still  `SIGCHILD` signal would be used for subprocesses, not every platforms  and kernel versions supports pidfd.

**Comment by gvanrossum** [comment]
> > So maybe the minimal solution would be to add something to signal to ask for the wakeup fd (if the OS allows it?) and then if a wakeup fd is already set, refrain from clobbering it,...
> 
> I think there is some confusion here, wakeup fd is not dependent on the OS, it is not some kernel space thing but rather a file descriptor to which ceval writes the signal number to wakeup from blocking select or other blocking calls on Python side.

You got me there. But nevertheless this means we could easily add a getter API so asyncio can avoid clobbering it.

**Comment by gvanrossum** [comment]
> > if a wakeup fd is already set, refrain from clobbering it, and instead disable the subprocess functionality of the event loop
> 
> That seems rather complicated for little gain, even if the entire child watcher machinery ought to be removed still `SIGCHILD` signal would be used for subprocesses, not every platform and kernel version supports pidfd.

Yes, it's little gain, but I think it's a legitimate bug that asyncio clobbers the wakeup fd when it is already in use by something else, and this is the smallest fix I can think of. (Maybe smaller still would be to pass an explicit flag to new_event_loop() that means "don't use wakeup fd", maybe the OP would be okay with that? @hidmic)

**Comment by gvanrossum** [comment]
The perverse thing here is that your use case is on Windows, which doesn't even have real signals.

There's one glimmer of hope -- `signal.set_wakeup_fd()` returns the old fd, so we can use it to raise that exception, like so:
```py
old_fd = signal.set_wakeup_fd(self._csock.fileno())
if old_fd != -1:
    signal.set_wakeup_fd(old_fd)
    raise OSError(...)
```
This has two flaws: it doesn't restore the `warn_on_full_buffer` flag, and it's possible that a signal arrived and wrote a byte to `fd` before we reset it. But it's better than we currently have.

**Comment by kumaraditya303** [comment]
A PR raising `RuntimeError` if wakeup fd is already set would be better than status quo. Adding multiple wakeup fd would require some discussion as it would requires changes to the eval loop and signal handling.

**Comment by gvanrossum** [comment]
Let's see a PR for the error if it's already set then. There could still be scenarios where this breaks "working" code, but maybe the breakage will do something useful (e.g. alert people that they have multiple subsystems fighting over the wakeup fd).

I agree that expanding the API to support multiple wakeup fds or multiple handlers would require more discussion -- for example what would we do to users of the old API?

FWIW the actual C code for the wakeup fd is a bit gruesome, I wouldn't want to be the one to make changes there (nearly everything's done one way for Windows and another way for UNIX).

**Comment by kumaraditya303** [comment]
@hidmic Interested in creating a PR for raising exception ?

**Comment by hidmic** [comment]
@kumaraditya303 sure

**Comment by hidmic** [comment]
@gvanrossum @kumaraditya303 see https://github.com/python/cpython/pull/96807

**Comment by gvanrossum** [comment]
@hidmic Alas, this broke the SSL tests, which are only run on a buildbot. Do you need help reproducing the failures?

**Comment by hidmic** [comment]
@gvanrossum I can reproduce. It's `ssltests` promoting warnings to exceptions (see [here](https://github.com/python/cpython/blob/aa671b48d62a00c5ab87fb102be15dd5eeac84dd/Lib/test/ssltests.py#L19)) and `asyncio` tests running into inconsistent signal handling setups over and over e.g. using a `signal.set_wakeup_fd` mock on proactor event loop construction but not on close.

I'll try to fix it.

**Comment by gvanrossum** [comment]
@hidmic Any progress on fixing this?

**Comment by willingc** [comment]
@hidmic or @kumaraditya303 Do either of you wish to make a PR for this issue? Thanks!

**Comment by willingc** [comment]
Next action: If there is no further activity on this issue by the September core dev sprint, let's review there and consider closing as stale.
